source/common/socket.cc

```diff
@@ -0,0 +1,95 @@
+#include "common/socket.h"
+
+#include "external/envoy/source/common/config/utility.h"
+
+namespace Nighthawk {
+
+class SocketConfigFactoryImpl : public UpstreamSocketConfigFactory {};
+
+Envoy::ProtobufTypes::MessagePtr UpstreamSocketConfigFactory::createEmptyConfigProto() {
+  return std::make_unique<nighthawk::transport_socket::TransportSocket>();
+}
+
+Envoy::Network::TransportSocketFactoryPtr UpstreamSocketConfigFactory::createTransportSocketFactory(
```

Comment:
I think this method needs an overview comment and also a comment that explains each line. I didn't understand the meaning of "inner" and "outer" at first, and we are dealing with both `auto& inner_config_factory` and `MessagePtr inner_factory_config`. 

Let me see if I followed this:

- We register `UpstreamSocketConfigFactory` with Envoy
- `UpstreamSocketConfigFactory` provides the name "NighthawkSocket" for Envoy's registry
- Envoy calls `UpstreamSocketConfigFactory::createTransportSocketFactory` when it needs a transport socket factory, passing the transport socket config `Message message` from the bootstrap
- We extract the "outer config" `nighthawk::transport_socket::TransportSocket` proto from `message`
- Contained within the outer config is the transport socket "inner config" proto the user specified, or a "raw buffer" transport socket config if they didn't specify it
- We look up the transport socket factory-factory of the inner config in the Envoy registry
- We create the inner transport socket factory, passing the inner config *again*
- We return a new outer transport socket factory that wraps the inner transport socket factory

Now that Envoy has the outer transport socket factory, Envoy will call it:
- the outer transport socket factory calls the inner transport socket factory to get the real transport socket
- the outer transport socket factory creates a `Socket`, passing the real transport socket
- `Socket` increments counters in some methods, and in every method delegates to the real transport socket